colossalai/zero/sharded_optim/sharded_optim_v2.py

```diff
@@ -52,20 +56,24 @@ def __init__(self,
                                              growth_interval=growth_interval,
                                              hysteresis=hysteresis,
                                              max_scale=max_scale)
-        self._found_overflow: Tensor = torch.FloatTensor([0]).to(self.device)
+        self._found_overflow: Tensor = torch.FloatTensor([0]).to(torch.cuda.current_device())
 
-        # Store fp32 params
+        # Store fp32 param shards
         self.master_params: Dict[Parameter, Tensor] = {}
 
         for group in optimizer.param_groups:
             for p in group['params']:
-                if hasattr(p, 'ca_attr'):
-                    assert p.ca_attr.is_sharded, 'ShardedAdam can be only used with sharded model'
-                    self.master_params[p] = p.ca_attr.payload(self.device)
-                else:
-                    self.master_params[p] = p.data.to(device=self.device)
-                if torch.is_floating_point(self.master_params[p]) and self.master_params[p].dtype != torch.float:
-                    self.master_params[p] = self.master_params[p].to(torch.float)
+                assert hasattr(p, 'col_attr'), 'The parameter must be wrapped with ShardedParam'
+                is_param_sharded = p.col_attr.data.is_sharded
+                if not is_param_sharded:
+                    # Param is no sharded
+                    # As we only store param shard, we shard it here
+                    self.shard_strategy.shard([p.col_attr.data])
+                self.master_params[p] = cast_tensor_to_fp32(p.col_attr.data.payload).to(self.device)
```

Comment:
`.to(self.device)` is not needed?

Reply:
I think you do not need to trigger time-consuming communication (in shard and gather). just view the local part of the fp16 payload?

Reply:
> `.to(self.device)` is not needed?

I think it influences the device of optimizer states, just in case fp16 param is not on CPU when we want to offload optimizer states.

Reply:
> I think you do not need to trigger time-consuming communication (in shard and gather). just view the local part of the fp16 payload?

Absolutely I can do this. However, in order to support more shard strategies in the future, I think this code snippet is more general.

Reply:
now you can just add a TODO(ver217) comments here

Reply:
Done